### PR TITLE
fix: stack-overflow caused by BQ recursion

### PR DESCRIPTION
**Pull request by Shahroz16**
closes: https://github.com/customerio/customerio-android/issues/248

### CODE REVIEW

File: sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt [hunk 1]
```diff
@@ -19,86 +18,67 @@ internal class QueueRunRequestImpl internal constructor(
...
+                    queueStorage.delete(taskStorageId)
+                }
 
-                        goToNextTask(inventory, totalNumberOfTasksToRun, lastFailedTask = nextTaskToRunInventoryItem)
+                result.isFailure -> {
+                    val error = result.exceptionOrNull()
+                    logger.debug("queue task $taskStorageId run failed $error")
+
+                    when (error as? CustomerIOError) {
+                        is CustomerIOError.HttpRequestsPaused, is CustomerIOError.NoHttpRequestMade -> {
```
Comment (Shahroz16) [inline sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt:60]: added `is CustomerIOError.NoHttpRequestMade` because if there is no connection, there is no point trying out every request and using resources, might as well just wait till next time.

**Review by levibostian — CHANGES_REQUESTED** [review]
So far, this looks like a good change, but I am not confident that it works yet. 

GitHub annotations highlights a couple places where there is code that does not get executed by tests (meaning possible missing tests) and some of the code in the loop I think isn't correct. 

This is a critical part of the code base, execution of the BQ. I strongly suggest that more tests are written for this `run` function to cover all the possible edge cases.

**Comment by Shahroz16** [comment]
@levibostian Thanks for looking into it,

> GitHub annotations highlights a couple places where there is code that does not get executed by tests (meaning possible missing tests)

The only highlight seems towards `log` statements it seems.  

> This is a critical part of the code base, execution of the BQ. I strongly suggest that more tests are written for this run function to cover all the possible edge cases.

Can you explain what specific tests are you looking for? asking because we didn't change any logic except move from recursion to iterative approach. So, what gave you confidence in the previous implementation with the current test suite, but now makes you feel less confident?

**Comment by levibostian** [comment]
@Shahroz16 Great questions. 

The latest test that you added for when a task cannot be found in storage was 1 missing test case. That's a great test case to add, thank you. 

Github annotations is telling me that there is 1 more missing branch for [this conditional statement](https://github.com/customerio/customerio-android/pull/251/files#diff-84815b58149a230f1d5cd0ea839c2d756898864fed99d8442358848fb8339b7eR29-R30). Adding a test for that scenario would be great. 

You do have a good point in that this is a refactor so existing tests should be all that we need. In my original comment I could have explained better that even though this is a refactor, I don't want to ship this code until the test suite is improved upon. If there are missing tests, I wanted to make sure that we *add tests* to the existing suite to feel more confident in the code. 

I recently found a missing test case in this exact same code block in the iOS SDK. When I added the missing test case, I actually found a bug. This makes me want to double check our test suite around this code in the Android SDK to make sure we trust it. I believe this PR is a good opportunity to do that. 

I reviewed the current test suite and it looks good to me except: 
* [this conditional statement](https://github.com/customerio/customerio-android/pull/251/files#diff-84815b58149a230f1d5cd0ea839c2d756898864fed99d8442358848fb8339b7eR29-R30) missing a test 
* Non-blocker for this PR, `QueueIntegrationTests` - could we add a test to reproduce this stackoverflow error? Then make sure that this refactor fixes the issue?

**Comment by Shahroz16** [comment]
@levibostian added more tests.

Regarding,
> could we add a test to reproduce this stackoverflow error? Then make sure that this refactor fixes the issue?

I did try, but wasn't able to do it because it depends on multiple factors like heap size and limiting it isn't that straight forward with unit tests.

**Comment by levibostian** [comment]
Tests are great! Thanks!

Oh, I just realized there is [1 more conversation to have resolved](https://github.com/customerio/customerio-android/pull/251#discussion_r1319777532). I do consider this a blocker for merging this PR.

**Comment by Shahroz16** [comment]
@levibostian just noticed my comment to it was never posted so did it again, it says `2 days ago` for me that's why I originally added it. Just making sure, you can see it now.

File: sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt (hunk 1 above)
Comment (Shahroz16) [reply sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt:63]: I don't think we should hinder getting improvement out unless we have a strong reason not to do so, it's a very small change but with a big performance improvement. I am quite comfortable making this change because `NoHttpRequestMade` is **only** being set when the network responds back with `null`. 

If you can kindly go through the code and express what makes you worry, please mention that I can respond to that concern but otherwise, renaming data types and iOS mirroring the same behavior shouldn't be the reason to block an improvement going out. The risk involved if this error shows up is it will skip an iteration unless the next time the connection goes through. 

I can request the other squad members to check out the update as well if that makes you feel more confident.

Comment (Shahroz16) [reply sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt:63]: @mrehan27 could you also go through this change when you get chance for more 👁️

Comment (mrehan27) [reply sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt:63]: I don't see any other usage for `NoHttpRequestMade` at the moment.  So the change looks good, no reason to block unless a we find a use case that fails and can be verified by tests.

**Review by mrehan27 — APPROVED** [review]

File: sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt (hunk 1 above)
Comment (Shahroz16) [reply sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt:63]: That's my point as well, this is a serious performance concern and i am comfortable with the check and the code backs it. But i would still add more tests for it to improve confidence in later releases.
